Improve `oml` output: align columns and add permission-based color highlighting

PuneetKumar1790 opened the pull request:
**Your checklist for this pull request**
- [x] I've read the guidelines for contributing to this repository.
- [x] I made sure to follow the project's coding style.
- [ ] I've documented every `RZ_API` function and struct this PR changes.
- [ ] I've added tests that prove my changes are effective (required for changes to `RZ_API`).
- [ ] I've updated the Rizin book with the relevant information (if needed).
- [x] I've used AI tools to generate fully or partially these code changes and I'm sure the changes are not copyrighted by somebody else.

**Detailed description**

This PR improves the readability of the `oml` command output by fixing column alignment and adding permission-based color highlighting.

The intent is to improve visual clarity while preserving existing behavior. All changes are localized and gated behind existing configuration options.

Changes included:
- Ensure consistent column alignment using formatted output
- Add permission-based color highlighting:
  - RWX → `widget_sel`
  - X → `ai_exec`
  - W → `ai_write`
  - R → `ai_read`
- Apply colors to:
  - start/end address columns
  - the visual block bar
  - the permission string
- Enable colors only when `scr.color > 0`
- Preserve exact output when color is disabled

All changes are localized to `rz_table_visual_list()` in `librz/util/table.c`.

No API or ABI changes are introduced.

**Test plan**

Manual testing was performed as this change affects interactive CLI output formatting.

Steps used to verify the change:
```sh
e scr.color=0; oml
e scr.color=1; oml
e scr.utf8=true; oml
e scr.utf8=false; oml
Verified:

Column alignment is correct in all modes

Permission-based colors match memory permissions

Visual block bar renders correctly

Output without colors matches previous behavior

No crashes or regressions observed

No automated tests were added, as this change affects terminal rendering and visual output only.

Closing issues

Closes #5704.

yaml
Copy code

notxvilka reviewed (changes requested):
Please follow the PR template

PuneetKumar1790 commented:
> Please follow the PR template

Thanks for the note I’ve updated the PR description to fully follow the template.

PuneetKumar1790 commented:
> Please follow the PR template

I reviewed CONTRIBUTING.md and added a command-based test under test/db/cmd/cmd_open to meet the new contributor test coverage requirement.
The test uses scr.color=0 and scr.utf8=false for deterministic output.

wargio reviewed (changes requested):
i'm not sure if this is the right way i would "colorify" tables.

PuneetKumar1790 commented:
> i'm not sure if this is the right way i would "colorify" tables.

Thanks for the feedback  that’s fair.

The intent was to improve readability while keeping the existing table structure intact, but I’m very open to adjusting the approach.

Could you clarify how you’d prefer table coloring to be handled?
For example, should coloring be applied at the table/column level rather than per-cell, or limited to specific fields only?

Happy to rework this to better match the intended style.

wargio commented:
tbh, i do not know. i think maybe having some apis to set the colors of each column could be ok, as this can become reusable.

@notxvilka

notxvilka commented:
> tbh, i do not know. i think maybe having some apis to set the colors of each column could be ok, as this can become reusable.
> 
> @notxvilka

Refactoring tables is way more complicated than this task. I would say let's colorize ad-hoc for now, [RzTable refactoring](https://github.com/rizinorg/rizin/issues/1961) can wait.

wargio commented:
> Refactoring tables is way more complicated than this task. I would say let's colorize ad-hoc for now, [RzTable refactoring](https://github.com/rizinorg/rizin/issues/1961) can wait.

Agreed, but i fear that this also messed up any non table output like the json and csv formatting

PuneetKumar1790 commented:
> > Refactoring tables is way more complicated than this task. I would say let's colorize ad-hoc for now, [RzTable refactoring](https://github.com/rizinorg/rizin/issues/1961) can wait.
> 
> Agreed, but i fear that this also messed up any non table output like the json and csv formatting

That’s a fair concern, thanks for pointing it out.

The changes are currently confined to `rz_table_visual_list()`, which is only used for the visual/interactive rendering of the `oml` output. JSON and CSV paths bypass this function and use their own serializers, so they shouldn’t be affected by the column formatting or color logic.

That said, I agree it’s important to be sure I can double-check the JSON/CSV output paths and adjust or guard the changes further if you’d prefer.

wargio commented:
ok, try then. fix the test and add screenshots in the first message

wargio commented:
also since you are at it, please extract `rz_table_visual_list` and move it in `ctable.c` (create a new file and fix the `meson.build` script) and move also the structure `RzListInfo` from the table headers.

rename the function to `rz_core_list_info_to_table()` and move its definition into the private header that is in `core/core_private.h`

(do not forget about `rz_listinfo_new` and `rz_listinfo_free`)

PuneetKumar1790 commented:
> also since you are at it, please extract `rz_table_visual_list` and move it in `ctable.c` (create a new file and fix the `meson.build` script) and move also the structure `RzListInfo` from the table headers.
> 
> rename the function to `rz_core_list_info_to_table()` and move its definition into the private header that is in `core/core_private.h`
> 
> (do not forget about `rz_listinfo_new` and `rz_listinfo_free`)
Got it, thanks.

I can fix the test and add screenshots to the PR description right away.

Regarding the refactoring (extracting `rz_table_visual_list`, moving `RzListInfo`, renaming to `rz_core_list_info_to_table`, and updating meson): this looks like a larger structural change touching core/table boundaries.

Would you prefer this to be done as part of this PR, or should we keep this PR focused on the `oml` output improvement and handle the refactor in a follow-up PR?
I’m fine either way, just want to make sure the scope is aligned before proceeding.

wargio commented:
just add the color changes

PuneetKumar1790 commented:
> just add the color changes

Got it, thanks for the clarification.

I’ll limit this PR to the ad-hoc color changes only and drop the alignment-related modifications. I’ll update the code and tests accordingly.

wargio commented:
please be aware that i moved the code and changed how the table stuff works. you can likely implement the color code via the new API.

PuneetKumar1790 committed:
feat: add permission-based coloring to oml using new table API

Use rz_table_set_color_selector() callback API instead of manual color
injection. This ensures colors only appear in visual output, not in JSON/CSV.

Changes:
- Add oml_table_color_selector() callback function
- Register color selector with rz_table_set_color_selector()
- Use exact permission matching (RZ_PERM_RWX, RZ_PERM_RW, RZ_PERM_RX)
- Remove manual ANSI color code injection from cell values

Color mappings:
- rwx -> widget_sel
- rw  -> ai_write
- rx  -> ai_exec
- r   -> ai_read

The callback-based approach follows the pattern from rz-diff.c and ensures
the table system automatically handles different output modes.

Rot127 commented on librz/core/cdebug.c:
```diff
@@ -1330,6 +1330,26 @@ RZ_IPI void rz_core_debug_signal_print(RzDebug *dbg, RzCmdStateOutput *state) {
+	}
+	RzConsPrintablePalette *pal = &cons->context->pal;
+	if ((perm & RZ_PERM_W) && (perm & RZ_PERM_X)) {
+		return pal->widget_sel;
```
Giving `wx` its own color doesn't make too much sense. Because these permissions are rarely, if at all use.
So add special colors for `rw`, `rwx`, `rx` instead.

PuneetKumar1790 replied:
> Giving `wx` its own color doesn't make too much sense. Because these permissions are rarely, if at all use. So add special colors for `rw`, `rwx`, `rx` instead.

Thanks for the feedback.

I’ve updated the permission color mapping to drop the special case for wx and instead handle the more common combinations:

rwx → widget_sel

rw → ai_write

rx → ai_exec

r → ai_read

The change is limited to the permission color helper in librz/core/cdebug.c and keeps the rest of the logic unchanged.

CI is re-running with the updated commit.

wargio reviewed (changes requested):
These changes does not follow what i have asked previously, use the new APIs.

Also having `(perm & RZ_PERM_R) && (perm & RZ_PERM_W) && (perm & RZ_PERM_X)` instead of just doing `perm & RZ_PERM_RWX` etc.. is kinda wrong.

Please start thinking with your head before pushing these changes, as the current ones make no sense.
AI agents are supposed to help you, not to do your thinking.

PuneetKumar1790 commented:
> These changes does not follow what i have asked previously, use the new APIs.
> 
> Also having `(perm & RZ_PERM_R) && (perm & RZ_PERM_W) && (perm & RZ_PERM_X)` instead of just doing `perm & RZ_PERM_RWX` etc.. is kinda wrong.
> 
> Please start thinking with your head before pushing these changes, as the current ones make no sense. AI agents are supposed to help you, not to do your thinking.

@wargio You’re right  sorry about the back and forth earlier. I hadn’t fully internalized the new table API, and that’s on me. I went back, dug into it properly, and reworked the implementation to align with the design you introduced.

This time I’m using the table color API the way it’s meant to be used. I added an oml_table_color_selector() callback, modeled after the patterns in rz-diff.c and cmd_api.c, and registered it with rz_table_set_color_selector(). I also removed all the manual color codes that were being pushed into the cell values.

For permissions, I switched to exact matching with (perm & RZ_PERM_RWX) == RZ_PERM_RWX, and the same approach for RW and RX, instead of checking individual bits.

As a result, the cells now only contain plain text like "rwx", and the table system handles coloring by calling the selector callback per cell. That also means JSON and CSV output stay clean automatically.

I went through commit ea21a4a7b4 where the table refactor happened and followed that architecture instead of trying to patch around it. In hindsight, I should’ve done that homework first instead of iterating blindly.

The changes are pushed and CI is running. If you still see anything off with the approach, let me know  happy to adjust.

notxvilka commented:
It violates our rule to use AI tools for issues specifically created to be as a starting point for new contributors because it *prevents understanding* which is the sole goal of such issues: https://github.com/rizinorg/rizin/blob/dev/CONTRIBUTING.md#getting-started

You can restart from scratch or take another issue and *do not use AI mindlessly*.
